homeassistant/components/isy994/sensor.py

```diff
@@ -46,48 +47,52 @@ class ISYSensorEntity(ISYNodeEntity):
     """Representation of an ISY994 sensor device."""
 
     @property
-    def raw_unit_of_measurement(self) -> str:
+    def raw_unit_of_measurement(self) -> Union[dict, str]:
         """Get the raw unit of measurement for the ISY994 sensor device."""
         uom = self._node.uom
 
         # Backwards compatibility for ISYv4 Firmware:
         if isinstance(uom, list):
             return UOM_FRIENDLY_NAME.get(uom[0], uom[0])
+
+        # Special cases for ISY UOM index units:
+        isy_states = UOM_TO_STATES.get(uom)
+        if isy_states:
+            return isy_states
+
         return UOM_FRIENDLY_NAME.get(uom)
 
     @property
     def state(self) -> str:
         """Get the state of the ISY994 sensor device."""
-        if self.value == ISY_VALUE_UNKNOWN:
-            return STATE_UNKNOWN
+        value = self._node.status
+        if value == ISY_VALUE_UNKNOWN:
+            return None
 
-        uom = self._node.uom
-        # Backwards compatibility for ISYv4 Firmware:
-        if isinstance(uom, list):
-            uom = uom[0]
-        if not uom:
-            return STATE_UNKNOWN
-
-        states = UOM_TO_STATES.get(uom)
-        if states and states.get(self.value):
-            return states.get(self.value)
-        if self._node.prec and int(self._node.prec) != 0:
-            str_val = str(self.value)
-            int_prec = int(self._node.prec)
-            decimal_part = str_val[-int_prec:]
-            whole_part = str_val[: len(str_val) - int_prec]
-            val = float(f"{whole_part}.{decimal_part}")
-            raw_units = self.raw_unit_of_measurement
-            if raw_units in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
-                val = self.hass.config.units.temperature(val, raw_units)
-            return val
-        return self.value
+        # Get the translated ISY Unit of Measurement
+        uom = self.raw_unit_of_measurement
+
+        # Check if this is a known index pair UOM
+        if isinstance(uom, dict):
+            return uom.get(value, value)
+
+        # Handle ISY precision and rounding
+        value = convert_isy_value_to_hass(value, uom, self._node.prec)
+
+        # Convert temperatures to Home Assistant's unit
+        if uom in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            value = self.hass.config.units.temperature(value, uom)
+
+        return value
 
     @property
     def unit_of_measurement(self) -> str:
-        """Get the unit of measurement for the ISY994 sensor device."""
+        """Get the Home Assistant unit of measurement for the device."""
         raw_units = self.raw_unit_of_measurement
-        if raw_units in (TEMP_FAHRENHEIT, TEMP_CELSIUS):
+        # Check if this is a known index pair UOM
+        if isinstance(raw_units, dict):
+            return None
+        if raw_units in (TEMP_FAHRENHEIT, TEMP_CELSIUS, UOM_DOUBLE_TEMP):
             return self.hass.config.units.temperature_unit
```

Comment:
How do we know that the device has the same temperature unit of measurement as the user has configured in home assistant?

Reply:
Unfortunately, we don't know for sure for most devices if they use `UOM_DOUBLE_TEMP` and this makes an assumption that they are the same. There's no endpoint that I've found to get the ISY's units.

For most thermostats that use this UOM, they have a dedicated `UOM` parameter set that we can use--which is what the `climate` platform uses [here](https://github.com/shbatm/home-assistant/blob/324b4c3725bdb50e5d7bd49805526bae6c44f6cd/homeassistant/components/isy994/climate.py#L108).

Reply:
Ok. But for the device that return just `TEMP_FAHRENHEIT` or `TEMP_CELSIUS`, wouldn't it be better to return what the device returns instead of returning what the user has configured. I'm thinking that a user with a Celsius configuration could use a device that uses Fahrenheit, and that case would cause a bug now.